source/slang/ir.cpp

```diff
@@ -471,6 +478,43 @@ namespace Slang
         return (IRBlock*)use->get();
     }
 
+    UInt IRUnconditionalBranch::getArgCount()
+    {
+        switch(op)
+        {
+        case kIROp_unconditionalBranch:
+            return getOperandCount() - 1;
+
+        case kIROp_loop:
+            return getOperandCount() - 3;
+
+        default:
+            SLANG_UNEXPECTED("unhandled unconditional branch opcode");
+            UNREACHABLE_RETURN(0);
+        }
+    }
+
+    IRUse* IRUnconditionalBranch::getArgs()
```

Comment:
It's not super clear to me what 'args' denotes here over 'operands'. Because of the nature of PR I'm assuming it's to do with the analysis of the flow of the use of variables, but it's not obvious.

Reply:
In the IR the terminology gets used as follows:

* Every instruction has zero or more *operands*, which are just that: the operands that would appear after the instruction mnemonic when we spit out assembly

* Some instructions (most notably function calls) have a subset of the operands that represent *arguments* being passed to a callee. So for a `call` instruction, operrand `0` is the function to call, and operands `1` and up are the arguments. So argument `0` is operand `1`

Our encoding of SSA "phi nodes" is that each basic block can have zero or more parameters, and when jumping to a basic block with a branch, the branch instruction passes arguments to those parameters. Similar to a call, a simple unconditional branch has operand `0` be the target to jump to, and operands `1` and up be the arguments to pass.

The way we encode our CFG we always break "critical edges" (edges from a block with multiple successors to one with multiple predecessors). The nature of SSA "phi nodes" is that they are only ever required on blocks with multiple predecessors (they need to be a at a join point), so we always know that a branch to such a block will be an unconditional branch. Thus our encoding only needs to pass arguments (the values for "phi nodes") when doing an unconditional jump.

Reply:
There is a strong point here that I need to do a write up on the IR encoding and how it structures the various invariants. We are using a non-standard encoding of SSA form that makes certain manipulations easier, but it means that stuff like this can't just be looked up in a compilers textbook.